src/main/java/com/demo/top/itunes/ItunesGateway.java

```diff
@@ -0,0 +1,61 @@
+package com.demo.top.itunes;
+
+import com.demo.top.exception.InternalServerException;
+import com.demo.top.model.album.AlbumSearchResponse;
+import com.demo.top.model.artist.ArtistSearchResponse;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.RestTemplate;
+
+@Component
+public class ItunesGateway {
+
+  private RestTemplate restTemplate;
+
+  private String itunesUrl;
+
+  private ObjectMapper objectMapper;
+
+  @Value("${top-application.itunes.url}")
+  public void setItunesUrl(String itunesUrl) {
+    this.itunesUrl = itunesUrl;
+  }
+
+
+  public ArtistSearchResponse getArtists(String artistName) {
+    ResponseEntity<String> responseEntity =  restTemplate.getForEntity(itunesUrl + "/search?entity=musicArtist&term=" + artistName, String.class);
+    return jsonToObject(responseEntity.getBody(),ArtistSearchResponse.class);
+  }
+
+  public AlbumSearchResponse getAlbums(Long amgArtistId, Integer limit) {
+    ResponseEntity<String> responseEntity =  restTemplate.getForEntity(itunesUrl + "/lookup?entity=album&amgArtistId=" + amgArtistId + "&limit=" + limit, String.class);
+    return jsonToObject(responseEntity.getBody(),AlbumSearchResponse.class);
+  }
+
+  private <T> T jsonToObject(String json, Class<T> clazz) {
+    if (json == null || json.length() == 0) {
+      throw new InternalServerException("no response body");
```

Comment:
If the response string is empty it's likely an error on itunes side (therefore not an internal error)

Reply:
Which error code do you suggest then?

Reply:
Changed to FailedDependencyException